File: src/python/bunkalunk/cache.py
"""HDF5 Decode Artifact IO Module."""

import os
from dataclasses import dataclass, fields, MISSING
from pathlib import Path
from tempfile import mkstemp

import h5py


CACHE_VERSION = 20260828
_SUFFIX = ".h5"


@dataclass
class CacheData:
    """This class strictly represents decode artifacts. It does not store
    operational state.

    """

    latitude: list[float | None]
    longitude: list[float | None]
    time: list[float]

    start_time: float
    sport: str | None = None
    heart_rate: list[float | None] | None = None
    elevation: list[float | None] | None = None
    distance: list[float | None] | None = None
    speed: list[float | None] | None = None


_CACHE_FIELDS_OPTIONAL_SCALAR = {"sport"}
# Optional row-aligned fields
_CACHE_FIELDS_OPTIONAL_VECTOR = {
    f.name
    for f in fields(CacheData)
    if f.default is not MISSING or f.default_factory is not MISSING
} - _CACHE_FIELDS_OPTIONAL_SCALAR


def _write_cache_file(file_path: Path, data: CacheData) -> None:
    # TODO: Refactor so that mandatory and optional fields are populated
    # dynamically, instead of hardcoding them
    n = len(data.time)
    if not (n == len(data.latitude) == len(data.longitude)):
        raise ValueError(
            "Unequal length coordinate arrays (time, latitude, longitude); "
            f"Got '{n}', '{data.latitude}', '{data.longitude}'"
        )

    if data.heart_rate and n != len(data.heart_rate):
        raise ValueError(f"Unequal length data array 'heart_rate'. Got '{n}'")

    with h5py.File(file_path, "w") as cache:
        time: h5py.Dataset = cache.create_dataset("time", shape=(n,), dtype="float64")
        latitude: h5py.Dataset = cache.create_dataset(
            "latitude", shape=(n,), dtype="float64"
        )
        longitude: h5py.Dataset = cache.create_dataset(
            "longitude", shape=(n,), dtype="float64"
        )

        cache.attrs.create("cache_version", CACHE_VERSION, dtype=int)

        cache.attrs.create("start_time", data.start_time, dtype=float)
        time[:] = data.time[:]
        latitude[:] = data.latitude[:]
        longitude[:] = data.longitude[:]

        for field_name in _CACHE_FIELDS_OPTIONAL_VECTOR:
            val = getattr(data, field_name, None)
            if val is not None:
                cache_var: h5py.Dataset = cache.create_dataset(
                    field_name, shape=(n,), dtype="float64"
                )
                cache_var[:] = val[:]

        # Only one optional scalar so no need to loop over the set
        if data.sport is not None:
            cache.attrs.create("sport", data.sport, dtype=h5py.string_dtype())


class CacheWriteFailure(Exception):
    pass


def write_cache(file_path: Path, data: CacheData) -> None:
    tmp, tmp_name = mkstemp(dir=file_path.parent, suffix=_SUFFIX)
    os.close(tmp)  # Make sure h5py.File gets a closed fd
    tmp_path = Path(tmp_name)
    try:
        _write_cache_file(tmp_path, data)
        tmp_path.replace(file_path)
    except Exception as e:
        tmp_path.unlink(missing_ok=True)
        raise CacheWriteFailure(f"Cache write for '{file_path.name}' failed.") from e


def resolve_cache_path(content_fingerprint: str, activity_store: Path) -> Path:
    shard = content_fingerprint[0:2]
    cache_parent = activity_store / shard
    if not cache_parent.exists():
        cache_parent.mkdir(parents=True, exist_ok=True)
    cache_path = (cache_parent / content_fingerprint).with_suffix(_SUFFIX)
    return cache_path

File: src/julia/src/cache.jl
using HDF5

const CACHE_VERSION = 20260701

struct CacheVersionMismatch <: Exception
    version::Int64
end

function Base.showerror(io::IO, e::CacheVersionMismatch)
    return print(io, "Expected cache version $CACHE_VERSION but got $(e.version)")
end

struct CacheData
    start_time::Float64
    time::Vector{Float64}
    latitude::Vector{Float64}
    longitude::Vector{Float64}
    sport::Union{String, Nothing}
    heart_rate::Union{Vector{Float64}, Nothing}
    elevation::Union{Vector{Float64}, Nothing}
    distance::Union{Vector{Float64}, Nothing}
    speed::Union{Vector{Float64}, Nothing}
end

const _CACHE_FIELDS_OPTIONAL = [
    fname for (fname, ftype) in zip(fieldnames(CacheData), fieldtypes(CacheData))
        if Nothing <: ftype
]

const _CACHE_FIELDS_OPTIONAL_SCALAR = [:sport]
const _CACHE_FIELDS_OPTIONAL_VECTOR = setdiff(
    _CACHE_FIELDS_OPTIONAL, _CACHE_FIELDS_OPTIONAL_SCALAR
)

CacheData(
    start_time,
    time,
    latitude,
    longitude
    ;
    sport = nothing,
    heart_rate = nothing,
    elevation = nothing,
    distance = nothing,
    speed = nothing
) = CacheData(
    start_time,
    time,
    latitude,
    longitude,
    sport,
    heart_rate,
    elevation,
    distance,
    speed
)

function _read_or_nothing(f::HDF5.File, name::String)
    return haskey(f, name) ? read(f, name) : nothing
end

function _read_or_nothing(f::HDF5.AttributeDict, name::String)
    return haskey(f, name) ? f[name] : nothing
end

function read_cache(file_path::String)
    return h5open(file_path, "r") do file
        file_attrs = attrs(file)

        let cache_version = file_attrs["cache_version"]
            cache_version == CACHE_VERSION || throw(
                CacheVersionMismatch(cache_version)
            )
        end

        CacheData(
            file_attrs["start_time"],
            read(file, "time"),
            read(file, "latitude"),
            read(file, "longitude")
            ;
            sport = _read_or_nothing(file_attrs, "sport"),
            heart_rate = _read_or_nothing(file, "heart_rate"),
            elevation = _read_or_nothing(file, "elevation"),
            speed = _read_or_nothing(file, "speed"),
            distance = _read_or_nothing(file, "distance")
        )
    end
end

find_valid_points(cache::CacheData) = @. !isnan(cache.latitude) & !isnan(cache.longitude)

function drop_cache_points(cache::CacheData, keep::BitVector)
    vector_kwargs = Dict{Symbol, Vector{<:Real}}()
    for field_name in _CACHE_FIELDS_OPTIONAL_VECTOR
        field_value = getfield(cache, field_name)
        if field_value !== nothing
            vector_kwargs[field_name] = field_value[keep]
        end
    end

    return CacheData(
        cache.start_time,
        cache.time[keep],
        cache.latitude[keep],
        cache.longitude[keep]
        ;
        sport = cache.sport,
        vector_kwargs...
    )
end
